### PR TITLE
Add/Improve push notification support

### PR DESCRIPTION
**Pull request by phillipthelen**
### Changes
- rename /user/addPushDevice to /user/add-push-device
- add API Call to remove an existent Push Device
- send push devices to iOS devices
- include more data in push notifications
- add options to enable/disable each push notification
- add push notifications settings to notification settings page
- add tests for the API calls

This PR still has one issue. Apple has an option to check for failed push devices. For some reason I have not yet been able to get one returned. I submit this PR now, so that it can be reviewed while I investigate this.

### CODE REVIEW

File: website/server/models/user/schema.js [hunk 1]
```diff
@@ -512,6 +523,7 @@ let schema = new Schema({
   pushDevices: [{
     regId: {type: String},
     type: {type: String},
+    createdAt: {type: Date, default: Date.now},
```
Comment (paglias) [inline website/server/models/user/schema.js]: for this it would be better to use a subschema with timestamps like we do for tags

File: website/server/models/group.js [hunk 2]
```diff
@@ -406,18 +406,24 @@ schema.methods.startQuest = async function startQuest (user) {
...
     });
     sendTxnEmail(membersToEmail, 'quest-started', [
       { name: 'PARTY_URL', content: '/#/options/groups/party' },
     ]);
+    let membersToPush = _.filter(membersToNotify, (member) => {
+      // send push notifications and filter users that disabled emails
+      return member.preferences.pushNotifications.questStarted !== false &&
+        member._id !== user._id;
+    });
+    _.each(membersToPush, function pushNotificationMembers (member) {
```
Comment (paglias) [inline website/server/models/group.js]: you should use an arrow function here

File: website/server/libs/api-v3/pushNotifications.js [hunk 3]
```diff
@@ -11,47 +12,58 @@ let gcm = GCM_API_KEY ? pushNotify.gcm({
...
+    console.log(devices);
   });
 }
 
-module.exports = function sendNotification (user, title, message, timeToLive = 15) {
-  return; // TODO push notifications are not currently enabled
+module.exports = function sendNotification (user, title, message, identifier, category = "", payload = {}, timeToLive = 15) {
 
   if (!user) return;
+  if (user.preferences && user.preferences.pushNotifications && user.preferences.pushNotifications.unsubscribeFromAll === true) return;
```
Comment (paglias) [inline website/server/libs/api-v3/pushNotifications.js]: you can just check for `user.preferences.pushNotifications.unsubscribeFromAll`, the parent objects will always exists

File: website/server/libs/api-v3/pushNotifications.js [hunk 4]
```diff
@@ -11,47 +12,58 @@ let gcm = GCM_API_KEY ? pushNotify.gcm({
...
-  });
+let apn = APN_KEY ? pushNotify.apn({
+  key: nconf.get('PUSH_CONFIGS:APN_PEM_FILES:KEY'),
+  cert: nconf.get('PUSH_CONFIGS:APN_PEM_FILES:CERT')
+}) : undefined;
 
-  gcm.on('transmissionError', (/* error, message, registrationId */) => {
-    // console.info("transmissionError", error, message, registrationId);
+if (apn) {
+  var feedback = apnLib.Feedback({
```
Comment (paglias) [inline website/server/libs/api-v3/pushNotifications.js]: let

File: website/server/libs/api-v3/pushNotifications.js [hunk 5]
```diff
@@ -11,47 +12,58 @@ let gcm = GCM_API_KEY ? pushNotify.gcm({
...
+    key: nconf.get('PUSH_CONFIGS:APN_PEM_FILES:KEY'),
+    cert: nconf.get('PUSH_CONFIGS:APN_PEM_FILES:CERT'),
+    "batchFeedback": true,
```
Comment (paglias) [inline website/server/libs/api-v3/pushNotifications.js]: no need for quotes

File: website/server/libs/api-v3/pushNotifications.js [hunk 6]
```diff
@@ -11,47 +12,58 @@ let gcm = GCM_API_KEY ? pushNotify.gcm({
...
+    "interval": 3600, //Check for feedback once an hour
   });
-
-  gcm.on('updated', (/* result, registrationId */) => {
-    // console.info("updated", result, registrationId);
+  feedback.on("feedback", function(devices) {
+    console.log("GOT FEEDBACK");
```
Comment (paglias) [inline website/server/libs/api-v3/pushNotifications.js]: having a console.log in the code will cause linting to fail

File: website/server/libs/api-v3/pushNotifications.js [hunk 7]
```diff
@@ -3,6 +3,7 @@
 import _ from 'lodash';
 import nconf from 'nconf';
 import pushNotify from 'push-notify';
+var apnLib = require('apn');
```
Comment (paglias) [inline website/server/libs/api-v3/pushNotifications.js]: import

File: website/server/controllers/api-v3/pushNotifications.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+ * @apiSuccess {string} message Success message
+ */
+api.addPushDevice = {
+  method: 'POST',
+  url: '/user/add-push-device',
+  middlewares: [authWithHeaders()],
+  async handler (req, res) {
+    let user = res.locals.user;
+
+    console.log("ASDF", req);
```
Comment (negue) [inline website/server/controllers/api-v3/pushNotifications.js]: can be removed ;)

File: website/server/libs/api-v3/pushNotifications.js [hunk 9]
```diff
@@ -11,47 +12,58 @@ let gcm = GCM_API_KEY ? pushNotify.gcm({
...
-    // console.info("updated", result, registrationId);
+  feedback.on("feedback", function(devices) {
+    console.log("GOT FEEDBACK");
+    console.log(devices);
   });
 }
 
-module.exports = function sendNotification (user, title, message, timeToLive = 15) {
-  return; // TODO push notifications are not currently enabled
+module.exports = function sendNotification (user, title, message, identifier, category = "", payload = {}, timeToLive = 15) {
```
Comment (paglias) [inline website/server/libs/api-v3/pushNotifications.js]: single quotes for category = ''

File: website/server/controllers/api-v3/quests.js [hunk 10]
```diff
@@ -106,11 +106,15 @@ api.inviteToQuest = {
...
-      sendPushNotification(
-        member,
-        common.i18n.t('questInvitationTitle'),
-        common.i18n.t('questInvitationInfo', { quest: quest.text() })
-      );
+      if (member.preferences.pushNotifications.invitedQuest !== false) {
+        sendPushNotification(
+          member,
+          common.i18n.t('questInvitationTitle'),
+          common.i18n.t('questInvitationInfo', {quest: quest.text()}),
```
Comment (paglias) [inline website/server/controllers/api-v3/quests.js]: quest.text should be passed `req.language`. In general where available you should use `res.t()` instead of `i18n.t` so the correct language will be applied

File: website/server/controllers/api-v3/pushNotifications.js [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+    if (indexOfPushDevice === -1) {
+      throw new BadRequest(res.t('pushDeviceNotFound', req.language));
+    }
+
+    pushDevices.splice(indexOfPushDevice);
+    user.pushDevices = pushDevices;
+    await user.save();
+
+    res.respond(200, ...[{pushDevices: user.pushDevices}, res.t('pushDeviceRemoved')]);
```
Comment (paglias) [inline website/server/controllers/api-v3/pushNotifications.js]: the spread operator here is not necessary (`...`) you can just pass the two arguments separately and not inside an array

File: website/server/controllers/api-v3/pushNotifications.js [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+      regId,
+    });
+
+
+    if (indexOfPushDevice === -1) {
+      throw new BadRequest(res.t('pushDeviceNotFound', req.language));
+    }
+
+    pushDevices.splice(indexOfPushDevice);
+    user.pushDevices = pushDevices;
```
Comment (paglias) [inline website/server/controllers/api-v3/pushNotifications.js]: no need to re-assign `user.pushDevices`, it's already modified by `splice`

File: website/server/controllers/api-v3/pushNotifications.js [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    let indexOfPushDevice = _.findIndex(pushDevices, {
+      regId,
+    });
+
+
+    if (indexOfPushDevice === -1) {
+      throw new BadRequest(res.t('pushDeviceNotFound', req.language));
+    }
+
+    pushDevices.splice(indexOfPushDevice);
```
Comment (paglias) [inline website/server/controllers/api-v3/pushNotifications.js]: you should pass `1` as the second parameter to remove 1 element

File: website/server/controllers/api-v3/pushNotifications.js [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+
+    let pushDevices = user.pushDevices;
+
+    let indexOfPushDevice = _.findIndex(pushDevices, {
+      regId,
+    });
+
+
+    if (indexOfPushDevice === -1) {
+      throw new BadRequest(res.t('pushDeviceNotFound', req.language));
```
Comment (paglias) [inline website/server/controllers/api-v3/pushNotifications.js]: res.t is passed automatically req.language

File: website/server/controllers/api-v3/pushNotifications.js [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+api.removePushDevice = {
+  method: 'GET',
+  url: '/user/remove-push-device/:regId',
+  middlewares: [authWithHeaders()],
+  async handler (req, res) {
+    let user = res.locals.user;
+
+    req.checkParams('regId', res.t('regIdRequired')).notEmpty();
+    let regId = req.params.regId;
+    if (!user.pushDevices) {
```
Comment (paglias) [inline website/server/controllers/api-v3/pushNotifications.js]: no need for this check, user.pushDevices will always exists

**Comment by crookedneighbor** [comment]
The `sendPushNotification` function is taking a lot of arguments. Would it be better for it to take an object so it's clearer what each arg is?

File: website/server/controllers/api-v3/pushNotifications.js [hunk 16]
```diff
@@ -0,0 +1,108 @@
+import { authWithHeaders } from '../../middlewares/api-v3/auth';
+import _ from 'lodash';
+import {
+  BadRequest,
+  NotAuthorized,
+} from '../../libs/api-v3/errors';
+
+let api = {};
+
```
Comment (paglias) [inline website/server/controllers/api-v3/pushNotifications.js:8]: I would scope these two routes as POST /user/push-devices and DELETE /user/push-devices/:regId

File: website/server/controllers/api-v3/groups.js [hunk 17]
```diff
@@ -537,11 +537,16 @@ async function _inviteByUUID (uuid, group, inviter, req, res) {
...
-    userToInvite,
-    common.i18n.t(group.type === 'guild' ? 'invitedGuild' : 'invitedParty'),
-    group.name
-  );
+  if (userToInvite.preferences.pushNotifications[`invited${groupLabel}`] !== false) {
+    let identifier = group.type === 'guild' ? 'invitedGuild' : 'invitedParty';
+    sendPushNotification(
+      userToInvite,
+      group.name,
+      common.i18n.t(identifier),
```
Comment (paglias) [inline website/server/controllers/api-v3/groups.js]: res.t

File: website/server/controllers/api-v3/pushNotifications.js (hunk 16 above)
Comment (crookedneighbor) [inline website/server/controllers/api-v3/pushNotifications.js:8]: Is there any case where we'd want to allow someone to remove a push device via an email link? That's the only case I can think for for using a GET.

**Comment by negue** [comment]
I did a mistake back then not adding a removeMethod so thats good @vIiRuS :+1:

**Comment by paglias** [comment]
I agree with @crookedneighbor about sendPushNotification taking an object instead of multiple arguments

**Comment by phillipthelen** [comment]
Should I remove all arguments and replace them with an object or just some of them?

File: common/locales/en/settings.json [hunk 18]
```diff
@@ -126,6 +128,7 @@
     "unsubscribedTextOthers": "You won't receive any other email from Habitica.",
     "unsubscribeAllEmails": "Check to Unsubscribe from Emails",
     "unsubscribeAllEmailsText": "By checking this box, I certify that I understand that by unsubscribing from all emails, Habitica will never be able to notify me via email about important changes to the site or my account.",
+    "unsubscribeAllPush": "Check to Unsubscribe from all Push Notifications",
```
Comment (crookedneighbor) [inline common/locales/en/settings.json:131]: Should `Unsubscribe` be capitalized? What about `Push Notifications`?

File: common/locales/en/settings.json [hunk 19]
```diff
@@ -138,7 +141,6 @@
     "promoCode": "Promo Code",
     "promoCodeApplied": "Promo Code Applied! Check your inventory",
     "promoPlaceholder": "Enter Promotion Code",
-    "couponText": "We sometimes have events and give out coupon codes for special gear. (eg, those who stop by our Wondercon booth)",
```
Comment (crookedneighbor) [inline common/locales/en/settings.json:141]: Why was this removed?

File: common/locales/en/settings.json [hunk 20]
```diff
@@ -152,8 +154,11 @@
     "invalidUrl": "invalid url",
     "invalidEnabled": "the \"enabled\" parameter should be a boolean",
     "regIdRequired": "RegId is required",
+    "invalidPushClient": "Invalid client. Only first-party Habitica clients can receive push notifications",
     "pushDeviceAdded": "Push device added successfully",
     "pushDeviceAlreadyAdded": "The user already has the push device",
+    "pushDeviceNotFound": "The user has no push device with this id.",
+    "pushDeviceRemoved": "Push device removed successfully",
```
Comment (crookedneighbor) [inline common/locales/en/settings.json]: End with a period?

File: common/locales/en/settings.json [hunk 21]
```diff
@@ -152,8 +154,11 @@
     "invalidUrl": "invalid url",
     "invalidEnabled": "the \"enabled\" parameter should be a boolean",
     "regIdRequired": "RegId is required",
+    "invalidPushClient": "Invalid client. Only first-party Habitica clients can receive push notifications",
```
Comment (crookedneighbor) [inline common/locales/en/settings.json]: Let's say "Official Habitica clients". And end with a period.

File: test/api/v3/integration/user/GET-user_remove_push_device.test.js [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+  it('returns an error if user does not have the push device', async () => {
+    await expect(user.get(`/user/remove-push-device/${regId}`))
+      .to.eventually.be.rejected.and.eql({
+        code: 400,
+        error: 'BadRequest',
+        message: t('pushDeviceNotFound'),
+      });
+  });
+
+// More tests in common code unit tests
```
Comment (crookedneighbor) [inline test/api/v3/integration/user/GET-user_remove_push_device.test.js]: This should go before the it blocks.

File: website/server/controllers/api-v3/pushNotifications.js [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+/**
+ * @apiIgnore
+ * @api {post} /api/v3/user/add-push-device Add a push device to a user
```
Comment (crookedneighbor) [inline website/server/controllers/api-v3/pushNotifications.js]: I think this could just be `/user/push-device` the POST implies that you are adding one.

Comment (crookedneighbor) [inline website/server/controllers/api-v3/pushNotifications.js]: I agree with @paglias though that it should be `push-devices` (plural)

File: website/server/controllers/api-v3/pushNotifications.js (hunk 16 above)
Comment (crookedneighbor) [inline website/server/controllers/api-v3/pushNotifications.js:8]: @vIiRuS can you confirm if that was your intention with the `GET`? If it was not, then it should definitely be a `DELETE`

File: website/server/controllers/api-v3/pushNotifications.js [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+ * @apiSuccess {string} message Success message
+ */
+api.addPushDevice = {
+  method: 'POST',
+  url: '/user/add-push-device',
+  middlewares: [authWithHeaders()],
+  async handler (req, res) {
+    let user = res.locals.user;
+
+    let regId = _.get(req, 'body.regId');
```
Comment (crookedneighbor) [inline website/server/controllers/api-v3/pushNotifications.js]: You should use the express validator checks for these. IE:

``` js
req.checkBody('regId', res.t('regIdRequired')).notEmpty();
req.checkBody('type', res.t('typeRequired')).notEmpty();

let validationErrors = req.validationErrors();
if (validationErrors) throw validationErrors;

let item = {
  regId: req.body.regId,
  type: req.body.type,
};
```

File: website/server/controllers/api-v3/pushNotifications.js [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+    if (!type) throw new BadRequest(res.t('typeRequired'));
+
+    let pushDevices = user.pushDevices;
+
+    let item = {
+      regId,
+      type,
+    };
+
+    let indexOfPushDevice = _.findIndex(pushDevices, {
```
Comment (crookedneighbor) [inline website/server/controllers/api-v3/pushNotifications.js]: How about:

``` js
if (pushDevices.find(device => device.regId === item.regId)) {
  throw new NotAuthorized(res.t('pushDeviceAlreadyAdded'));
}
```

Comment (crookedneighbor) [inline website/server/controllers/api-v3/pushNotifications.js]: Alternatively, we could use the mongo's `$addToSet` and not worry about duplicates.

File: website/server/controllers/api-v3/pushNotifications.js [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+/**
+ * @apiIgnore
+ * @api {get} /api/v3/user/remove-push-device remove a push device from a user
+ * @apiVersion 3.0.0
+ * @apiName UserRemovePushDevice
+ * @apiGroup User
+ *
+ * @apiParam {string} regId The id of the push device
+ *
+ * @apiSuccess {Object} data List of push devices
```
Comment (crookedneighbor) [inline website/server/controllers/api-v3/pushNotifications.js:63]: If it's for a specific id, shouldn't it just return the one?

Comment (crookedneighbor) [inline website/server/controllers/api-v3/pushNotifications.js:63]: Oh, sorry, forgot this was to delete the devices.

File: website/server/controllers/api-v3/pushNotifications.js (hunk 16 above)
Comment (crookedneighbor) [inline website/server/controllers/api-v3/pushNotifications.js:8]: Also, should there be a `GET` method for fetching all your push devices?

File: website/server/controllers/api-v3/pushNotifications.js [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+ * @apiSuccess {string} message Success message
+ */
+api.removePushDevice = {
+  method: 'GET',
+  url: '/user/remove-push-device/:regId',
+  middlewares: [authWithHeaders()],
+  async handler (req, res) {
+    let user = res.locals.user;
+
+    req.checkParams('regId', res.t('regIdRequired')).notEmpty();
```
Comment (crookedneighbor) [inline website/server/controllers/api-v3/pushNotifications.js:73]: This doesn't validate automatically, you need to use 

``` js
let validationErrors = req.validationErrors();
if (validationErrors) throw validationErrors;
```

File: website/server/controllers/api-v3/pushNotifications.js [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+    let regId = req.params.regId;
+
+    let pushDevices = user.pushDevices;
+
+    let indexOfPushDevice = _.findIndex(pushDevices, {
```
Comment (crookedneighbor) [inline website/server/controllers/api-v3/pushNotifications.js]: Preference for native `findIndex` method on Arrays.

File: website/server/controllers/api-v3/pushNotifications.js [hunk 29]
```diff
@@ -0,0 +1,99 @@
...
+      regId,
+    });
+
+
```
Comment (crookedneighbor) [inline website/server/controllers/api-v3/pushNotifications.js]: Extra space here.

File: website/server/libs/api-v3/pushNotifications.js [hunk 30]
```diff
@@ -11,47 +12,58 @@ let gcm = GCM_API_KEY ? pushNotify.gcm({
...
-    // console.info("updated", result, registrationId);
+  feedback.on("feedback", function(devices) {
+    //console.log("GOT FEEDBACK");
+    //console.log(devices);
   });
 }
 
-module.exports = function sendNotification (user, title, message, timeToLive = 15) {
-  return; // TODO push notifications are not currently enabled
+module.exports = function sendNotification (user, title, message, identifier, category = '', payload = {}, timeToLive = 15) {
```
Comment (crookedneighbor) [inline website/server/libs/api-v3/pushNotifications.js]: I think the function should probably be:

``` js
function sendNotification (user, details = {}) {
```

Comment (crookedneighbor) [inline website/server/libs/api-v3/pushNotifications.js]: or options = {} or payload = {}

And then you could have private method that converts the payload to ios and android formats.

**Comment by crookedneighbor** [comment]
@vIiRuS Let me know when you're ready for me to look at this again. Overall, good work so far 👍

File: common/locales/en/settings.json (hunk 18 above)
Comment (khipkin) [inline common/locales/en/settings.json:131]: I agree that those words should start with lowercase letters.

File: common/locales/en/settings.json [hunk 31]
```diff
@@ -108,10 +108,12 @@
     "emailNotifications": "Email Notifications",
     "wonChallenge": "You won a Challenge!",
     "newPM": "Received Private Message",
+    "newPMInfo": "New Message from <%= name %>: <%= message %>",
     "sentGems": "Sent gems!",
     "giftedGems": "Gifted Gems",
-    "giftedGemsInfo": "<%= amount %> Gems - by <%= name %>",
+    "giftedGemsInfo": "<%= name %> gifted you <%= amount %> Gems",
     "giftedSubscription": "Gifted Subscription",
+    "giftedSubscriptionInfo": "<%= name %> gifted you a <%= months %> Subscription",
```
Comment (khipkin) [inline common/locales/en/settings.json:116]: Some of these changes seem unrelated to the rest of the PR.  Specifically, the changes regarding the `{newPM,giftedGems,giftedSubscription}Info` strings.

File: common/script/ops/addPushDevice.js [hunk 32]
```diff
@@ -1,41 +1 @@
-import _ from 'lodash';
```
Comment (khipkin) [inline common/script/ops/addPushDevice.js:1]: Why keep this file?

File: common/locales/en/settings.json (hunk 31 above)
Comment (khipkin) [inline common/locales/en/settings.json:116]: Ah nevermind, I see where the strings are used.  But maybe the string changes/additions should still be a separate (mini) PR for clarity.

Comment (phillipthelen) [inline common/locales/en/settings.json:116]: The strings are the ones used for the content of the push notifications. Separating those into a separate PR doesn't seem to make much sense, since the one PR wouldn't work without the other.

File: common/locales/en/settings.json (hunk 18 above)
Comment (phillipthelen) [inline common/locales/en/settings.json:131]: The capitalization is the same as with the unsubscribeAllEmails string. So it should probably be also changed there, if it's changed here?

File: common/locales/en/settings.json (hunk 19 above)
Comment (phillipthelen) [inline common/locales/en/settings.json:141]: Wasn't intentional. Might have been accidentally through git merging.

File: website/server/controllers/api-v3/pushNotifications.js (hunk 16 above)
Comment (phillipthelen) [inline website/server/controllers/api-v3/pushNotifications.js:8]: No, it wasn't really my intention to allow it to be included in an email. I'll change it to DELETE.
I wouldn't need such an API call for the official clients. But if others think having a GET method to list all devices would make sense, I can add it.

File: common/locales/en/settings.json (hunk 31 above)
Comment (khipkin) [inline common/locales/en/settings.json:116]: Sorry, I meant the addition of the new push notifications as well as the related string changes.  But then one PR would be dependent on the other, so that may still be undesirable.

**Comment by phillipthelen** [comment]
@crookedneighbor fixed all the issues!

File: common/locales/en/settings.json (hunk 19 above)
Comment (paglias) [inline common/locales/en/settings.json:141]: It's added back at line 145

File: common/script/ops/addPushDevice.js (hunk 32 above)
Comment (paglias) [inline common/script/ops/addPushDevice.js:1]: This file isn't keep, it's been removed

File: test/api/v3/integration/user/DELETE-user_push_device.test.js [hunk 33]
```diff
@@ -0,0 +1,34 @@
...
+describe('DELETE /user/push-devices', () => {
+  let user;
+  let regId = '10';
+  let type = 'ios';
+
+  beforeEach(async () => {
+    user = await generateUser();
+  });
+
+  // More tests in common code unit tests
```
Comment (paglias) [inline test/api/v3/integration/user/DELETE-user_push_device.test.js]: this comment should be removed

File: test/api/v3/integration/user/DELETE-user_push_device.test.js [hunk 34]
```diff
@@ -0,0 +1,34 @@
...
+
+  it('returns an error if user does not have the push device', async () => {
+    await expect(user.del(`/user/push-devices/${regId}`))
+      .to.eventually.be.rejected.and.eql({
+        code: 400,
```
Comment (paglias) [inline test/api/v3/integration/user/DELETE-user_push_device.test.js]: should be a 404 not a 400

File: website/server/controllers/api-v3/pushNotifications.js (hunk 23 above)
Comment (paglias) [inline website/server/controllers/api-v3/pushNotifications.js]: The docs here must be updated

File: website/server/controllers/api-v3/pushNotifications.js [hunk 35]
```diff
@@ -0,0 +1,95 @@
...
+    if (pushDevices.find(device => device.regId === item.regId)) {
+      throw new NotAuthorized(res.t('pushDeviceAlreadyAdded'));
+    }
+
+    pushDevices.push(item);
+
+    await user.save();
+
+    res.respond(200, {pushDevices: user.pushDevices}, res.t('pushDeviceAdded'));
```
Comment (paglias) [inline website/server/controllers/api-v3/pushNotifications.js]: maybe here pass user.pushDevices directly instead of wrapping them inside an object. This way you'll get {data: {pushDevices: []}}, isn't better {data: []}?

File: website/server/controllers/api-v3/pushNotifications.js [hunk 36]
```diff
@@ -0,0 +1,95 @@
...
+  },
+};
+
+/**
+ * @apiIgnore
+ * @api {get} /api/v3/user/remove-push-device remove a push device from a user
```
Comment (paglias) [inline website/server/controllers/api-v3/pushNotifications.js]: the url here should be updated

File: website/server/libs/api-v3/pushNotifications.js [hunk 37]
```diff
@@ -11,47 +12,62 @@ let gcm = GCM_API_KEY ? pushNotify.gcm({
...
+  let feedback = apnLib.Feedback({
+    key: nconf.get('PUSH_CONFIGS:APN_PEM_FILES:KEY'),
+    cert: nconf.get('PUSH_CONFIGS:APN_PEM_FILES:CERT'),
+    batchFeedback: true,
+    interval: 3600, //Check for feedback once an hour
   });
-
-  gcm.on('updated', (/* result, registrationId */) => {
-    // console.info("updated", result, registrationId);
+  feedback.on("feedback", function(devices) {
```
Comment (paglias) [inline website/server/libs/api-v3/pushNotifications.js]: single quotes and arrow function

File: website/server/libs/api-v3/pushNotifications.js [hunk 38]
```diff
@@ -11,47 +12,62 @@ let gcm = GCM_API_KEY ? pushNotify.gcm({
...
+    //console.log("GOT FEEDBACK");
+    //console.log(devices);
   });
 }
 
-module.exports = function sendNotification (user, title, message, timeToLive = 15) {
-  return; // TODO push notifications are not currently enabled
-
+module.exports = function sendNotification (user, details={}) {
```
Comment (paglias) [inline website/server/libs/api-v3/pushNotifications.js]: spaces `details = {}`

File: website/server/libs/api-v3/pushNotifications.js [hunk 39]
```diff
@@ -11,47 +12,62 @@ let gcm = GCM_API_KEY ? pushNotify.gcm({
...
         }
-
         break;
 
       case 'ios':
+        if (apn) {
+          apn.send({
+            token: pushDevice.regId,
+            alert: details.message,
+            sound: "default",
```
Comment (paglias) [inline website/server/libs/api-v3/pushNotifications.js]: single quotes

File: website/server/models/pushDevice.js [hunk 40]
```diff
@@ -0,0 +1,21 @@
...
+  regId: {type: String, required: true},
+  type: {type: String},
+}, {
+  strict: true,
+  minimize: false, // So empty objects are returned
+  _id: false, // use id instead of _id
+});
+
+schema.plugin(baseModel, {
+  noSet: ['_id'],
```
Comment (paglias) [inline website/server/models/pushDevice.js]: when using `id` instead of `_id` you should also include `id` inside `noSet`

**Comment by paglias** [comment]
@vIiRuS a thing I just noticed is that the push notifications lib is excluded from linting because it was not ported to v3, can you removed the exception? https://github.com/HabitRPG/habitrpg/blob/develop/website/server/libs/api-v3/pushNotifications.js#L1

**Comment by phillipthelen** [comment]
@paglias also made the changes you noted

File: website/server/libs/api-v3/pushNotifications.js [hunk 41]
```diff
@@ -11,47 +10,62 @@ let gcm = GCM_API_KEY ? pushNotify.gcm({
   retries: 3,
 }) : undefined;
 
-// TODO review and test this file when push notifications are added back
+let APN_KEY = nconf.get('PUSH_CONFIGS:APN_PEM_FILES:KEY');
```
Comment (paglias) [inline website/server/libs/api-v3/pushNotifications.js]: const

File: website/server/libs/api-v3/pushNotifications.js [hunk 42]
```diff
@@ -11,47 +10,62 @@ let gcm = GCM_API_KEY ? pushNotify.gcm({
...
 
-if (gcm) {
-  gcm.on('transmitted', (/* result, message, registrationId */) => {
-    // console.info("transmitted", result, message, registrationId);
-  });
+let apn = APN_KEY ? pushNotify.apn({
+  key: nconf.get('PUSH_CONFIGS:APN_PEM_FILES:KEY'),
```
Comment (paglias) [inline website/server/libs/api-v3/pushNotifications.js]: you already load the variable at line 13

File: website/server/libs/api-v3/pushNotifications.js [hunk 43]
```diff
@@ -11,47 +10,62 @@ let gcm = GCM_API_KEY ? pushNotify.gcm({
...
+  cert: nconf.get('PUSH_CONFIGS:APN_PEM_FILES:CERT'),
+}) : undefined;
 
-  gcm.on('transmissionError', (/* error, message, registrationId */) => {
-    // console.info("transmissionError", error, message, registrationId);
+if (apn) {
+  let feedback = new apnLib.Feedback({
```
Comment (paglias) [inline website/server/libs/api-v3/pushNotifications.js:23]: is it supposed here to use both `apnLib` and `pushNotify.apn`?

Comment (phillipthelen) [inline website/server/libs/api-v3/pushNotifications.js:23]: yes. pushNotify doesn't provide access to the feedback service.

**Comment by paglias** [comment]
See https://github.com/HabitRPG/habitrpg/pull/7682
